### PR TITLE
Slightly improve performance of hydration.

### PR DESCRIPTION
**Pull request by bmeurer**
Avoid loading nodeType and data couple times from the same node in a row,
but instead load them only once, which will help engines to run this code
faster, especially during startup of the application. The general approach
is still not ideal, since hydrating this way forces the browser engine
to materialize JavaScript wrapper objects for all DOM nodes, even if they
are not interesting to hydration itself.

### CODE REVIEW

**Comment by sebmarkbage** [comment]
In normal usage, it is not expected that there will be any non-hydratable nodes. So these would all be materialized anyway. This is here mostly as a precaution against browser extensions or proxies inserting unexpected content.

**Comment by bmeurer** [comment]
Thanks. I'm wondering if it's possible to do something about this in general. While it's probably possible to optimize this a bit more with the current approach, having to materialize all these wrapper objects is always gonna remain a bottleneck.

cc @developit @ timneutkens

**Comment by sebmarkbage** [comment]
One thing we could do is drop all validation but still preresolve references to the node. Is there a way to walk the DOM while not materializing JS wrappers or cheaper ones?

We used to have a lazy approach but that instead paid the cost when the user interacted with something and also added complexity and meta-data that made it overall not worth it.

We’re moving to a selective hydration approach driven by interaction which effectively makes the paths lazy but also able to warm up in preparation for an interaction. So the characteristics are about to change.

**Comment by bmeurer** [comment]
Interesting. The lazy approach sounds like a viable path.

Not validating, i.e. not loading `nodeType` and doing these lower case comparisons can probably reduce worst case scenarios - where we right not have easily 600ms for hydration - significantly.

File: packages/react-dom/src/client/ReactDOMHostConfig.js [hunk 1]
```diff
@@ -670,44 +670,40 @@ export function registerSuspenseInstanceRetry(
...
-  ) {
-    node = node.nextSibling;
+  for (; node != null; node = node.nextSibling) {
+    const nodeType = node.nodeType;
+    if (nodeType === ELEMENT_NODE || nodeType === TEXT_NODE) {
+      break;
+    }
+    if (enableSuspenseServerRenderer) {
+      if (nodeType === COMMENT_NODE) {
+        break;
```
Comment (sebmarkbage) [inline packages/react-dom/src/client/ReactDOMHostConfig.js]: This should actually have been `continue`. Logic is hard. :)

Comment (acdlite) [reply packages/react-dom/src/client/ReactDOMHostConfig.js]: Also, test coverage is hard :D

**Pull request merged**